dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/ModuleSelectionConfigImpl.java

```diff
@@ -23,72 +23,136 @@
+            serverxmlmode = false;
+        } else {
+            serverxmlmode = configFromServerXML.getIncludedModules().isPresent() || configFromServerXML.getExcludedModules().isPresent();
+        }
+
+        if (serverxmlmode) {
+            //In another PR we will change first to a version specific service
+            inclusion = configFromServerXML.getIncludedModules().orElse("first").trim();
+        } else {
+            inclusion = configFromMPConfig.getOptionalValue(Constants.MERGE_INCLUDE_CONFIG, String.class).orElse("first");
+        }
 
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            Tr.debug(null, tc, "Names in config: " + config.getPropertyNames());
+            if (serverxmlmode) {
+                Tr.debug(null, tc, "Acquired includes statement from server.xml");
+            } else {
+                Tr.debug(null, tc, "Names in config: " + configFromMPConfig.getPropertyNames());
+            }
             Tr.debug(null, tc, "Inclusion read from config: " + inclusion);
         }
 
         if (inclusion.equals("none")) {
-            result.included = Collections.emptyList();
+            included = Collections.emptyList();
         } else if (inclusion.equals("all")) {
-            result.isAll = true;
+            isAll = true;
         } else if (inclusion.equals("first")) {
-            result.isFirst = true;
+            isFirst = true;
+        } else {
+            included = parseModuleNames(inclusion, Constants.MERGE_INCLUDE_CONFIG);
+        }
+
+        String exclusion;
+
+        if (serverxmlmode) {
+            exclusion = configFromServerXML.getExcludedModules().orElse("none").trim();
         } else {
-            result.included = parseModuleNames(inclusion, Constants.MERGE_INCLUDE_CONFIG);
+            exclusion = configFromMPConfig.getOptionalValue(Constants.MERGE_EXCLUDE_CONFIG, String.class).orElse("none")
+                                          .trim();
         }
 
-        String exclusion = config.getOptionalValue(Constants.MERGE_EXCLUDE_CONFIG, String.class).orElse("none").trim();
         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
             Tr.debug(null, tc, "Exclusion read from config: " + exclusion);
         }
 
         if (exclusion.equals("none")) {
-            result.excluded = Collections.emptyList();
+            excluded = Collections.emptyList();
         } else {
-            result.excluded = parseModuleNames(exclusion, Constants.MERGE_EXCLUDE_CONFIG);
+            excluded = parseModuleNames(exclusion, Constants.MERGE_EXCLUDE_CONFIG);
         }
 
-        return result;
+        hasBeenInit = true;
+        if (TraceComponent.isAnyTracingEnabled() && tc.isEntryEnabled()) {
+            Tr.exit(tc, "lazyInit", toString());
+        }
     }
 
     @Override
     public String toString() {
+        lazyInit();
         StringBuilder sb = new StringBuilder("Module Selection Config[");
         if (isFirst) {
```

Comment:
I'm a bit concerned about `toString()` doing initialization here. That's the sort of thing that can cause enabling trace to change behaviour...

Could we make it so that `toString()` reports "Uninitialized Config` or something similar if initialization hasn't happened yet?